### PR TITLE
This is a reissue of a previous PR with a UT race condition fix.

### PR DESCRIPTION
**Pull request by archisgore**
This adds the optional capability to SyncCluster with etcd (described in detail in  the commit message.)

The bigger change which took me some time to figure out was the race condition causing UTs to fail occasionally (described in detail in the other commit message.)

I believe with the UT fix, two of the other pending PRs should pass as well.

### CODE REVIEW

File: engine/etcdng/etcd.go [hunk 1]
```diff
@@ -729,6 +733,13 @@ func (n *ng) deleteKey(key string) error {
 	return convertErr(err)
 }
 
+func (n *ng) periodicallySyncCluster() {
+	for {
```
Comment (klizhentas) [inline engine/etcdng/etcd.go]: this goroutine should exit when `Close` has been called. I suggest adding close channel and doing

```
select {
     case time.After <-...
      case <- n.closeC <-...
```

**Comment by archisgore** [comment]
Thanks for the feedback Alexander. Does this look good to you?

**Comment by klizhentas** [comment]
spent some time about `restartC` panic on race condition. The fact that we need to guard it with the mutex seems a bit wrong - after all, channels are sync mechanism. How about instead, we will make buffered channel for restartC (e.g. 100)- and do not close it at all? in this case, when stop will be called, we can make sure that no panic will occur and the goroutine would not block too.

**Comment by archisgore** [comment]
That was my original idea too. When I don't close the channel restartC, the UT hangs. You can try it for yourself if you check out head to my commit, and then remove the part where I close restartC.

I don't know the code well enough to know who/what is waiting on it. I know the sole listener on that channel does break successfully though. I'm happy to investigate a bit later.

**Comment by klizhentas** [comment]
oh, it's been a while since I wrote this chunk, you are right, here's what is waiting:

https://github.com/mailgun/vulcand/blob/master/supervisor/supervisor.go#L240

**Comment by klizhentas** [comment]
ok, in this case we can do the following:

* intro new channel - `broadcastCloseC` - it will be used to signal all interested parties that we are about to shutdown. We will only close it in `Stop`
* `supervise` will be changed to select for both `broadcastCloseC` and `restartC`:

```go
select {
   case <- restartC: // do real restart
   case <- broadcastCloseC: // call stop()
}
```
this will let `supervise` to exit without blocking on `restartC`. What do you think?

**Comment by archisgore** [comment]
Yup. Sounds simple enough. I'll have this done tonight or early tomorrow. I hate locking too. I love this approach. Thanks for the suggestion!

**Comment by klizhentas** [comment]
@archisgore thanks for spending your time on this problem!

**Comment by archisgore** [comment]
I think this should do it. I introduced the same situation by adding delays and it works every time consistently. No more locks!

**Comment by archisgore** [comment]
Pinging for reminder again. This change too I've been running in production for over two weeks and works very reliably.

**Comment by klizhentas** [comment]
@archisgore sorry for the delay. One last comment - let's make `broadcastC` a buffered channel to avoid the case when multiple goroutines created in `init` can be potentially blocked on the channel forever once `supervise` returns

**Comment by archisgore** [comment]
You got it! Coming up in 10 minutes.

**Comment by archisgore** [comment]
It's done. All checks passed!

**Comment by klizhentas** [comment]
lgtm, thanks for your efforts!

**Pull request merged**